File: lib/features/clan/models/clan_history.dart
enum ClanLeaderboardType {
  homeVillage('clan_home_points'),
  builderBase('clan_builder_base_points'),
  clanCapital('clan_capital_points');

  const ClanLeaderboardType(this.apiValue);

  final String apiValue;
}

class ClanLeaderboardHistory {
  const ClanLeaderboardHistory({required this.items});

  final List<ClanLeaderboardHistoryEntry> items;

  factory ClanLeaderboardHistory.fromJson(Map<String, dynamic> json) {
    return ClanLeaderboardHistory(
      items: _maps(
        json['items'],
      ).map(ClanLeaderboardHistoryEntry.fromJson).toList(growable: false),
    );
  }
}

class ClanLeaderboardHistorySummary {
  const ClanLeaderboardHistorySummary({required this.seasons});

  final List<ClanLeaderboardSeasonSummary> seasons;

  factory ClanLeaderboardHistorySummary.fromJson(Map<String, dynamic> json) {
    return ClanLeaderboardHistorySummary(
      seasons: _maps(
        json['seasons'],
      ).map(ClanLeaderboardSeasonSummary.fromJson).toList(growable: false),
    );
  }
}

class ClanLeaderboardSeasonSummary {
  const ClanLeaderboardSeasonSummary({
    required this.season,
    required this.after,
    required this.before,
    required this.daysInTop200,
    required this.bestRank,
    required this.peakPoints,
  });

  final String season;
  final DateTime after;
  final DateTime before;
  final int daysInTop200;
  final int bestRank;
  final int peakPoints;

  factory ClanLeaderboardSeasonSummary.fromJson(Map<String, dynamic> json) {
    return ClanLeaderboardSeasonSummary(
      season: json['season']?.toString() ?? '',
      after: _date(json['after']),
      before: _date(json['before']),
      daysInTop200: _integer(json['daysInTop200']),
      bestRank: _integer(json['bestRank']),
      peakPoints: _integer(json['peakPoints']),
    );
  }
}

class ClanLeaderboardHistoryEntry {
  const ClanLeaderboardHistoryEntry({
    required this.date,
    required this.rank,
    required this.points,
    required this.members,
    this.location,
  });

  final DateTime date;
  final int rank;
  final int points;
  final int members;
  final ClanLeaderboardLocation? location;

  factory ClanLeaderboardHistoryEntry.fromJson(Map<String, dynamic> json) {
    final location = json['location'];
    return ClanLeaderboardHistoryEntry(
      date: _date(json['date']),
      rank: _integer(json['rank']),
      points: _integer(
        json['clanPoints'] ??
            json['builderBasePoints'] ??
            json['capitalPoints'],
      ),
      members: _integer(json['members']),
      location: location is Map
          ? ClanLeaderboardLocation.fromJson(
              Map<String, dynamic>.from(location),
            )
          : null,
    );
  }
}

class ClanLeaderboardLocation {
  const ClanLeaderboardLocation({
    required this.id,
    required this.name,
    required this.isCountry,
    this.countryCode,
  });

  final int id;
  final String name;
  final bool isCountry;
  final String? countryCode;

  factory ClanLeaderboardLocation.fromJson(Map<String, dynamic> json) {
    return ClanLeaderboardLocation(
      id: _integer(json['id']),
      name: json['name']?.toString() ?? '',
      isCountry: json['isCountry'] == true,
      countryCode: json['countryCode']?.toString(),
    );
  }
}

class ClanLegendHistory {
  const ClanLegendHistory({required this.items});

  final List<ClanLegendHistoryEntry> items;

  factory ClanLegendHistory.fromJson(Map<String, dynamic> json) {
    return ClanLegendHistory(
      items: _maps(
        json['items'],
      ).map(ClanLegendHistoryEntry.fromJson).toList(growable: false),
    );
  }
}

class ClanLegendHistorySummary {
  const ClanLegendHistorySummary({
    required this.seasons,
    required this.topFinishes,
  });

  final List<ClanLegendSeasonSummary> seasons;
  final List<ClanLegendHistoryEntry> topFinishes;

  factory ClanLegendHistorySummary.fromJson(Map<String, dynamic> json) {
    return ClanLegendHistorySummary(
      seasons: _maps(
        json['seasons'],
      ).map(ClanLegendSeasonSummary.fromJson).toList(growable: false),
      topFinishes: _maps(
        json['topFinishes'],
      ).map(ClanLegendHistoryEntry.fromJson).toList(growable: false),
    );
  }
}

class ClanLegendSeasonSummary {
  const ClanLegendSeasonSummary({
    required this.season,
    required this.after,
    required this.before,
    required this.playerCount,
  });

  final String season;
  final DateTime after;
  final DateTime before;
  final int playerCount;

  factory ClanLegendSeasonSummary.fromJson(Map<String, dynamic> json) {
    return ClanLegendSeasonSummary(
      season: json['season']?.toString() ?? '',
      after: _date(json['after']),
      before: _date(json['before']),
      playerCount: _integer(json['playerCount']),
    );
  }
}

class ClanLegendHistoryEntry {
  const ClanLegendHistoryEntry({
    required this.season,
    required this.tag,
    required this.name,
    required this.expLevel,
    required this.trophies,
    required this.attackWins,
    required this.defenseWins,
    required this.rank,
  });

  final String season;
  final String tag;
  final String name;
  final int expLevel;
  final int trophies;
  final int attackWins;
  final int defenseWins;
  final int rank;

  factory ClanLegendHistoryEntry.fromJson(Map<String, dynamic> json) {
    return ClanLegendHistoryEntry(
      season: json['season']?.toString() ?? '',
      tag: json['tag']?.toString() ?? '',
      name: json['name']?.toString() ?? '',
      expLevel: _integer(json['expLevel']),
      trophies: _integer(json['trophies']),
      attackWins: _integer(json['attackWins']),
      defenseWins: _integer(json['defenseWins']),
      rank: _integer(json['rank']),
    );
  }
}

class ClanRecords {
  const ClanRecords({this.clanPoints, this.warWinStreak});

  final ClanRecord? clanPoints;
  final ClanRecord? warWinStreak;

  bool get isEmpty => clanPoints == null && warWinStreak == null;

  factory ClanRecords.fromJson(Map<String, dynamic> json) {
    return ClanRecords(
      clanPoints: _record(json['clanPoints']),
      warWinStreak: _record(json['warWinStreak']),
    );
  }
}

class ClanRecord {
  const ClanRecord({required this.value, required this.time});

  final int value;
  final DateTime time;

  factory ClanRecord.fromJson(Map<String, dynamic> json) {
    return ClanRecord(
      value: _integer(json['value']),
      time: _date(json['time']),
    );
  }
}

class ClanProfileHistory {
  const ClanProfileHistory({required this.items});

  final List<ClanProfileChange> items;

  factory ClanProfileHistory.fromJson(Map<String, dynamic> json) {
    return ClanProfileHistory(
      items: _maps(
        json['items'],
      ).map(ClanProfileChange.fromJson).toList(growable: false),
    );
  }
}

enum ClanProfileChangeType { clanLevel, description, unknown }

class ClanProfileChange {
  const ClanProfileChange({
    required this.time,
    required this.type,
    required this.previous,
    required this.current,
  });

  final DateTime time;
  final ClanProfileChangeType type;
  final Object? previous;
  final Object? current;

  factory ClanProfileChange.fromJson(Map<String, dynamic> json) {
    return ClanProfileChange(
      time: _date(json['time']),
      type: switch (json['type']?.toString()) {
        'clanLevel' => ClanProfileChangeType.clanLevel,
        'description' => ClanProfileChangeType.description,
        _ => ClanProfileChangeType.unknown,
      },
      previous: json['previous'],
      current: json['current'],
    );
  }
}

ClanRecord? _record(Object? value) {
  if (value is! Map) return null;
  return ClanRecord.fromJson(Map<String, dynamic>.from(value));
}

Iterable<Map<String, dynamic>> _maps(Object? value) sync* {
  if (value is! List) return;
  for (final item in value) {
    if (item is Map) yield Map<String, dynamic>.from(item);
  }
}

int _integer(Object? value) =>
    value is num ? value.toInt() : int.tryParse(value?.toString() ?? '') ?? 0;

DateTime _date(Object? value) =>
    DateTime.tryParse(value?.toString() ?? '')?.toUtc() ??
    DateTime.fromMillisecondsSinceEpoch(0, isUtc: true);

File: expo/src/features/clan/models/clan-history.ts
import {
  bool,
  dateOrEpoch,
  nullableString,
  record,
  records,
  string,
  type JsonRecord,
} from './parsing';

export const ClanLeaderboardType = {
  homeVillage: 'clan_home_points',
  builderBase: 'clan_builder_base_points',
  clanCapital: 'clan_capital_points',
} as const;
export type ClanLeaderboardTypeValue =
  (typeof ClanLeaderboardType)[keyof typeof ClanLeaderboardType];

export class ClanLeaderboardLocation {
  constructor(
    readonly id: number,
    readonly name: string,
    readonly isCountry: boolean,
    readonly countryCode: string | null,
  ) {}
  static fromJson(json: JsonRecord) {
    return new ClanLeaderboardLocation(
      historyInt(json.id),
      string(json.name),
      bool(json.isCountry),
      nullableString(json.countryCode),
    );
  }
}

export class ClanLeaderboardHistoryEntry {
  constructor(
    readonly date: Date,
    readonly rank: number,
    readonly points: number,
    readonly members: number,
    readonly location: ClanLeaderboardLocation | null,
  ) {}
  static fromJson(json: JsonRecord) {
    const location = record(json.location);
    return new ClanLeaderboardHistoryEntry(
      dateOrEpoch(json.date),
      historyInt(json.rank),
      historyInt(json.clanPoints ?? json.builderBasePoints ?? json.capitalPoints),
      historyInt(json.members),
      Object.keys(location).length ? ClanLeaderboardLocation.fromJson(location) : null,
    );
  }
}

export class ClanLeaderboardHistory {
  constructor(readonly items: readonly ClanLeaderboardHistoryEntry[]) {}
  static fromJson(json: JsonRecord) {
    return new ClanLeaderboardHistory(
      records(json.items).map(ClanLeaderboardHistoryEntry.fromJson),
    );
  }
}

export class ClanLeaderboardSeasonSummary {
  constructor(
    readonly season: string,
    readonly after: Date,
    readonly before: Date,
    readonly daysInTop200: number,
    readonly bestRank: number,
    readonly peakPoints: number,
  ) {}
  static fromJson(json: JsonRecord) {
    return new ClanLeaderboardSeasonSummary(
      string(json.season),
      dateOrEpoch(json.after),
      dateOrEpoch(json.before),
      historyInt(json.daysInTop200),
      historyInt(json.bestRank),
      historyInt(json.peakPoints),
    );
  }
}

export class ClanLeaderboardHistorySummary {
  constructor(readonly seasons: readonly ClanLeaderboardSeasonSummary[]) {}
  static fromJson(json: JsonRecord) {
    return new ClanLeaderboardHistorySummary(
      records(json.seasons).map(ClanLeaderboardSeasonSummary.fromJson),
    );
  }
}

export class ClanLegendHistoryEntry {
  constructor(
    readonly season: string,
    readonly tag: string,
    readonly name: string,
    readonly expLevel: number,
    readonly trophies: number,
    readonly attackWins: number,
    readonly defenseWins: number,
    readonly rank: number,
  ) {}
  static fromJson(json: JsonRecord) {
    return new ClanLegendHistoryEntry(
      string(json.season),
      string(json.tag),
      string(json.name),
      historyInt(json.expLevel),
      historyInt(json.trophies),
      historyInt(json.attackWins),
      historyInt(json.defenseWins),
      historyInt(json.rank),
    );
  }
}

export class ClanLegendHistory {
  constructor(readonly items: readonly ClanLegendHistoryEntry[]) {}
  static fromJson(json: JsonRecord) {
    return new ClanLegendHistory(records(json.items).map(ClanLegendHistoryEntry.fromJson));
  }
}

export class ClanLegendSeasonSummary {
  constructor(
    readonly season: string,
    readonly after: Date,
    readonly before: Date,
    readonly playerCount: number,
  ) {}
  static fromJson(json: JsonRecord) {
    return new ClanLegendSeasonSummary(
      string(json.season),
      dateOrEpoch(json.after),
      dateOrEpoch(json.before),
      historyInt(json.playerCount),
    );
  }
}

export class ClanLegendHistorySummary {
  constructor(
    readonly seasons: readonly ClanLegendSeasonSummary[],
    readonly topFinishes: readonly ClanLegendHistoryEntry[],
  ) {}
  static fromJson(json: JsonRecord) {
    return new ClanLegendHistorySummary(
      records(json.seasons).map(ClanLegendSeasonSummary.fromJson),
      records(json.topFinishes).map(ClanLegendHistoryEntry.fromJson),
    );
  }
}

export class ClanRecord {
  constructor(
    readonly value: number,
    readonly time: Date,
  ) {}
  static fromJson(json: JsonRecord) {
    return new ClanRecord(historyInt(json.value), dateOrEpoch(json.time));
  }
}

export class ClanRecords {
  constructor(
    readonly clanPoints: ClanRecord | null,
    readonly warWinStreak: ClanRecord | null,
  ) {}
  static fromJson(json: JsonRecord) {
    const points = record(json.clanPoints);
    const streak = record(json.warWinStreak);
    return new ClanRecords(
      Object.keys(points).length ? ClanRecord.fromJson(points) : null,
      Object.keys(streak).length ? ClanRecord.fromJson(streak) : null,
    );
  }
  get isEmpty() {
    return this.clanPoints === null && this.warWinStreak === null;
  }
}

export type ClanProfileChangeType = 'clanLevel' | 'description' | 'unknown';

export class ClanProfileChange {
  constructor(
    readonly time: Date,
    readonly type: ClanProfileChangeType,
    readonly previous: unknown,
    readonly current: unknown,
  ) {}
  static fromJson(json: JsonRecord) {
    const type = string(json.type);
    return new ClanProfileChange(
      dateOrEpoch(json.time),
      type === 'clanLevel' || type === 'description' ? type : 'unknown',
      json.previous,
      json.current,
    );
  }
}

export class ClanProfileHistory {
  constructor(readonly items: readonly ClanProfileChange[]) {}
  static fromJson(json: JsonRecord) {
    return new ClanProfileHistory(records(json.items).map(ClanProfileChange.fromJson));
  }
}

export function clanLeaderboardApiValue(type: ClanLeaderboardTypeValue): string {
  return type;
}

function historyInt(value: unknown): number {
  if (typeof value === 'number' && Number.isFinite(value)) return Math.trunc(value);
  const parsed = Number.parseInt(String(value ?? ''), 10);
  return Number.isNaN(parsed) ? 0 : parsed;
}
